### PR TITLE
Merge dataframe and dataframe imp

### LINKED ISSUE #1962: Merge Dataframe trait and DataframeImpl into Dataframe struct
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Follow up for https://github.com/apache/arrow-datafusion/pull/1712#issuecomment-1025266914.

The Dataframe trait was introduced in order to have two separate Dataframe implementation in Datafusion and Ballista. Since then, the design has changed and we are sharing the same dataframe implementation between Datafusion and Ballista now. Therefore, we could simplify the code base by getting rid of the Dataframe trait abstraction and promoting DataframeImpl into Dataframe struct.

**Describe the solution you'd like**

Dataframe trait and DataframeImpl merged into a single Dataframe struct

**Additional context**

Should help unblock https://github.com/apache/arrow-datafusion/pull/1712

### PR DESCRIPTION
**Pull request by vchag**
# Which issue does this PR close?

Closes #1962 .

 # Rationale for this change
The Dataframe trait was introduced in order to have two separate Dataframe implementation in Datafusion and Ballista. Since then, the design has changed and we are sharing the same dataframe implementation between Datafusion and Ballista now. Therefore, we could simplify the code base by getting rid of the Dataframe trait abstraction and promoting DataframeImpl into Dataframe struct.

# What changes are included in this PR?

- Merging Dataframe trait and DataframeImpl into a single Dataframe struct
- Deleting datafusion/src/execution/dataframe_impl.rs

### CODE REVIEW

**Review by xudong963 — COMMENTED** [review]
LGTM, thank you @vchag 

Btw, some examples in  `datafusion-examples/examples` still use dataframe impl.

**Comment by vchag** [comment]
> LGTM, thank you @vchag
> 
> Btw, some examples in `datafusion-examples/examples` still use dataframe impl.

You're welcome @xudong963 
Thanks for pointing it out. 

The log mentions of an 
unresolved import `datafusion::execution::dataframe_impl` 
in  datafusion-examples/examples/custom_datasource.rs:24:28

`datafusion-examples/examples/custom_datasource.rs` isn't present in 
datafusion-examples.

Could the build process be using an old version of datafusion-examples module?

File: datafusion/src/dataframe.rs [hunk 1]
```diff
@@ -229,25 +325,28 @@ pub trait DataFrame: Send + Sync {
...
-    /// # use datafusion::prelude::*;
-    /// # use datafusion::error::Result;
-    /// # #[tokio::main]
-    /// # async fn main() -> Result<()> {
-    /// let mut ctx = ExecutionContext::new();
-    /// let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new()).await?;
-    /// let batches = df.collect().await?;
-    /// # Ok(())
-    /// # }
-    /// ```
```
[the remark this answers is not in the document]
Comment (vchag) [reply datafusion/src/dataframe.rs]: Surething, revised the dataframe.rs files as suggested.

**Comment by xudong963** [comment]
> `datafusion-examples/examples/custom_datasource.rs` isn't present in
> datafusion-examples.

It exists in master
image


However, the ticket didn't delete it
image

**Comment by vchag** [comment]
> > `datafusion-examples/examples/custom_datasource.rs` isn't present in
> > datafusion-examples.
> 
> It exists in master image
> 
> However, the ticket didn't delete it image
> 
> strange...

Sorry, my bad. My branch was lagging. Merging apache: master into my branch fixed the issue.

**Review by xudong963 — APPROVED** [review]
Thanks again @vchag

**Review by yjshen — APPROVED** [review]
Thanks @vchag! Very neat refactoring!

**Comment by yjshen** [comment]
Thanks @vchag @xudong963 !

**Comment by vchag** [comment]
> Thanks @vchag @xudong963 !

yw!

**Comment by houqp** [comment]
nice work @vchag !

**Pull request merged**